include/swift/AST/ASTContext.h

```diff
@@ -296,6 +296,11 @@ class ASTContext final {
   mutable llvm::DenseMap<unsigned, Identifier>
     CanonicalGenericTypeParamTypeNames;
 
+  /// Maps canonical generic paramenter types to their non-canonical source
+  /// (useful for diagnostics).
+  mutable llvm::DenseMap<std::pair<unsigned, unsigned>, GenericTypeParamType *>
```

Comment:
It doesn't make sense for this mapping to be global, because multiple generic parameters with different names might canonicalize to the same type

Reply:
Ah I see, thanks @slavapestov  
I'm thinking of going for an approach where we would have special canonicalization for diagnostics where we skip canonicalize `GenericTypeParamType` as @xedin suggested. 
I was thinking of maybe store the reference for the origin declaration inside the canonical `GenericTypeParamType` that way we won't have the risk of it accumulating information (since is just a reference for it), and we could remap it in an easier way on diagnostics. But I'm definitely not sure this would be a good idea... does that make sense?

Reply:
The problem is that multiple non-canonical GenericTypeParamTypes can map to the same canonical GenericTypeParamType, so storing the original decl doesn't work. If it did, we wouldn't need the canonical types in the first place.

As @xedin suggested, you might want to try skipping the canonicalization altogether, and instead walking the Type and desugaring TypeAliasTypes while leaving other sugared types such as GenericTypeParamTypes alone.

Reply:
Ah sorry, I requested the review but forgot to respond ... 

> The problem is that multiple non-canonical GenericTypeParamTypes can map to the same canonical GenericTypeParamType, so storing the original decl doesn't work. If it did, we wouldn't need the canonical types in the first place.

Ah make sense thanks!

> As @xedin suggested, you might want to try skipping the canonicalization altogether, and instead walking the Type and desugaring TypeAliasTypes while leaving other sugared types such as GenericTypeParamTypes alone.

I made the changes, let me know if the implementation matches the idea :)